#!/usr/bin/env python3
"""gRPC OCR ì„œë¹„ìŠ¤ í…ŒìŠ¤íŠ¸ ìŠ¤í¬ë¦½íŠ¸

ì‚¬ìš©ë²•:
    python grpc_test.py health
    python grpc_test.py extract
    python grpc_test.py batch
"""

import asyncio
import sys
from pathlib import Path
from typing import Optional

# í”„ë¡œì íŠ¸ ë£¨íŠ¸ë¥¼ Python pathì— ì¶”ê°€
sys.path.insert(0, str(Path(__file__).parent / "packages" / "shared"))
sys.path.insert(0, str(Path(__file__).parent / "packages" / "celery_worker"))

import grpc
from shared.grpc.generated import common_pb2, ocr_pb2, ocr_pb2_grpc


class GrpcTester:
    """gRPC í…ŒìŠ¤íŠ¸ í´ë¼ì´ì–¸íŠ¸"""

    def __init__(self, server_address: str = "localhost:50051"):
        self.server_address = server_address
        self.channel = None
        self.stub = None

    async def connect(self):
        """gRPC ì„œë²„ ì—°ê²°"""
        self.channel = grpc.aio.insecure_channel(
            self.server_address,
            options=[
                ('grpc.max_send_message_length', 100 * 1024 * 1024),
                ('grpc.max_receive_message_length', 100 * 1024 * 1024),
            ]
        )
        self.stub = ocr_pb2_grpc.OCRServiceStub(self.channel)
        print(f"âœ… ì—°ê²°ë¨: {self.server_address}")

    async def close(self):
        """ì—°ê²° ì¢…ë£Œ"""
        if self.channel:
            await self.channel.close()
            print("ğŸ”Œ ì—°ê²° ì¢…ë£Œ")

    async def test_health(self):
        """í—¬ìŠ¤ ì²´í¬ í…ŒìŠ¤íŠ¸"""
        print("\n=== í—¬ìŠ¤ ì²´í¬ í…ŒìŠ¤íŠ¸ ===")

        request = ocr_pb2.HealthCheckRequest(
            service_name="OCRService"
        )

        try:
            if self.stub is None :
                raise Exception("self.stub is None")

            response = await self.stub.CheckHealth(request)

            print(f"ğŸ“Š ìƒíƒœ: {common_pb2.Status.Name(response.status)}")
            print(f"ğŸ”§ ì—”ì§„: {response.engine_type}")
            print(f"âœ“ ëª¨ë¸ ë¡œë“œ: {response.model_loaded}")
            print(f"ğŸ“Œ ë²„ì „: {response.version}")

            return response.status == common_pb2.STATUS_SUCCESS

        except grpc.RpcError as e:
            print(f"âŒ ì˜¤ë¥˜: {e.code()} - {e.details()}")
            return False

    async def test_extract_text(
        self,
        private_path: str = "/data/test.jpg",
        public_path: str = "/test/test.jpg"
    ):
        """OCR í…ìŠ¤íŠ¸ ì¶”ì¶œ í…ŒìŠ¤íŠ¸"""
        print("\n=== OCR í…ìŠ¤íŠ¸ ì¶”ì¶œ í…ŒìŠ¤íŠ¸ ===")
        print(f"ğŸ“‚ ì´ë¯¸ì§€: {private_path}")

        request = ocr_pb2.OCRRequest(
            public_image_path=public_path,
            private_image_path=private_path,
            language="korean",
            confidence_threshold=0.5,
            use_angle_cls=True
        )

        try:
            if self.stub is None :
                raise Exception()

            response = await self.stub.ExtractText(request)

            print(f"\nğŸ“Š ìƒíƒœ: {common_pb2.Status.Name(response.status)}")
            print(f"ğŸ“ ì „ì²´ í…ìŠ¤íŠ¸: {response.text[:100]}...")
            print(f"ğŸ¯ ì „ì²´ ì‹ ë¢°ë„: {response.overall_confidence:.2f}")
            print(f"ğŸ“¦ í…ìŠ¤íŠ¸ ë°•ìŠ¤ ìˆ˜: {len(response.text_boxes)}")

            # í…ìŠ¤íŠ¸ ë°•ìŠ¤ ìƒì„¸ ì •ë³´
            for idx, box in enumerate(response.text_boxes[:5]):  # ìµœëŒ€ 5ê°œë§Œ í‘œì‹œ
                print(f"\n  [{idx + 1}] í…ìŠ¤íŠ¸: {box.text}")
                print(f"      ì‹ ë¢°ë„: {box.confidence:.2f}")
                print(f"      ì¢Œí‘œ: {list(box.bbox.coordinates)[:4]}...")

            if len(response.text_boxes) > 5:
                print(f"\n  ... ì™¸ {len(response.text_boxes) - 5}ê°œ")

            # ë©”íƒ€ë°ì´í„°
            if response.metadata.data:
                print("ğŸ“‹ ë©”íƒ€ë°ì´í„°:")
                for key, value in response.metadata.data.items():
                    print(f"  - {key}: {value}")

            # ì—ëŸ¬ í™•ì¸
            if response.status == common_pb2.STATUS_FAILURE:
                print("âŒ ì—ëŸ¬:")
                print(f"  ì½”ë“œ: {response.error.code}")
                print(f"  ë©”ì‹œì§€: {response.error.message}")
                print(f"  ìƒì„¸: {response.error.details}")

            return response.status == common_pb2.STATUS_SUCCESS

        except grpc.RpcError as e:
            print(f"âŒ gRPC ì˜¤ë¥˜: {e.code()} - {e.details()}")
            return False

    async def test_batch_extract(
        self,
        image_paths: Optional[list[tuple[str, str]]] = None
    ):
        """ë°°ì¹˜ OCR í…ŒìŠ¤íŠ¸ (Server Streaming)"""
        print("\n=== ë°°ì¹˜ OCR í…ŒìŠ¤íŠ¸ ===")

        if image_paths is None:
            image_paths = [
                ("/data/test1.jpg", "/test/test1.jpg"),
                ("/data/test2.jpg", "/test/test2.jpg"),
                ("/data/test3.jpg", "/test/test3.jpg"),
            ]

        print(f"ğŸ“‚ ì´ë¯¸ì§€ ìˆ˜: {len(image_paths)}")

        # ìš”ì²­ ìƒì„±
        paths = [
            ocr_pb2.ImagePath(private_path=priv, public_path=pub)
            for priv, pub in image_paths
        ]

        request = ocr_pb2.OCRBatchRequest(
            image_paths=paths,
            language="korean",
            confidence_threshold=0.5,
            use_angle_cls=True
        )

        try:
            if self.stub is None :
                raise Exception()
            # ìŠ¤íŠ¸ë¦¬ë° ì‘ë‹µ ìˆ˜ì‹ 
            async for progress in self.stub.ExtractTextBatch(request):
                print(f"\nğŸ“Š ì§„í–‰ë¥ : {progress.progress_percentage:.1f}%")
                print(f"   ì²˜ë¦¬: {progress.processed_images}/{progress.total_images}")

                # í˜„ì¬ ê²°ê³¼
                result = progress.current_result
                print(f"   ìƒíƒœ: {common_pb2.Status.Name(result.status)}")
                print(f"   í…ìŠ¤íŠ¸ ë°•ìŠ¤: {len(result.text_boxes)}ê°œ")
                print(f"   ì‹ ë¢°ë„: {result.overall_confidence:.2f}")

            print("\nâœ… ë°°ì¹˜ ì²˜ë¦¬ ì™„ë£Œ!")
            return True

        except grpc.RpcError as e:
            print(f"âŒ gRPC ì˜¤ë¥˜: {e.code()} - {e.details()}")
            return False


async def main():
    """ë©”ì¸ í•¨ìˆ˜"""
    import argparse

    parser = argparse.ArgumentParser(description="gRPC OCR í…ŒìŠ¤íŠ¸")
    parser.add_argument(
        "command",
        choices=["health", "extract", "batch", "all"],
        help="ì‹¤í–‰í•  í…ŒìŠ¤íŠ¸"
    )
    parser.add_argument(
        "--server",
        default="localhost:50051",
        help="gRPC ì„œë²„ ì£¼ì†Œ (ê¸°ë³¸ê°’: localhost:50051)"
    )
    parser.add_argument(
        "--image",
        default="/data/test.jpg",
        help="í…ŒìŠ¤íŠ¸ ì´ë¯¸ì§€ ê²½ë¡œ"
    )

    args = parser.parse_args()

    tester = GrpcTester(args.server)

    try:
        await tester.connect()

        if args.command == "health":
            await tester.test_health()

        elif args.command == "extract":
            await tester.test_extract_text(
                private_path=args.image,
                public_path="/test/test.jpg"
            )

        elif args.command == "batch":
            await tester.test_batch_extract()

        elif args.command == "all":
            print("ğŸš€ ì „ì²´ í…ŒìŠ¤íŠ¸ ì‹œì‘\n")
            success = True

            success &= await tester.test_health()
            success &= await tester.test_extract_text()
            success &= await tester.test_batch_extract()

            print("\n" + "=" * 50)
            if success:
                print("âœ… ëª¨ë“  í…ŒìŠ¤íŠ¸ í†µê³¼!")
            else:
                print("âŒ ì¼ë¶€ í…ŒìŠ¤íŠ¸ ì‹¤íŒ¨")
            print("=" * 50)

    except Exception as e:
        print(f"âŒ ì˜ˆì™¸ ë°œìƒ: {str(e)}")
        import traceback
        traceback.print_exc()

    finally:
        await tester.close()


if __name__ == "__main__":
    asyncio.run(main())
